lib/src/rules/avoid_library_directive.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2015, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc =
+    'Avoid library directives unless that have documentation comments or '
+    'annotations';
+
+const _details = r'''
+**DO** use library directives if you want to document a library and/or annotate 
+a library.
+
+**GOOD:**
+```dart
+/// This library does important things
+library;
+```
+
+```dart
+@TestOn('js')
+library;
+```
+
+**BAD:**
+```dart
+library;
+```
+''';
+
+class AvoidLibraryDirective extends LintRule {
+  static const LintCode code = LintCode('avoid_library_directive',
+      'Library directives without comments or annotations should be avoided.',
+      correctionMessage: 'Try deleting the library directive.');
+
+  AvoidLibraryDirective()
+      : super(
+            name: 'avoid_library_directive',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addLibraryDirective(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitLibraryDirective(LibraryDirective node) {
+    if (node.documentationComment == null &&
+        node.sortedCommentAndAnnotations.isEmpty) {
```

Comment:
We should not report a lint for library directives in libraries that contain part directives.

Reply:
How would one do that?

Reply:
The parent of the `LibraryDirective` is a `CompilationUnit`, and from there you can get to all the other `directives` to see whether there are any part directives.

If you only want to lint when at least one of the parts uses the library name, that will be a bit harder.

Also, as per the issue, we might want to not lint when the library directive includes a name (though I'm not sure whether we need to care about `dart:mirrors` uses at this point).

Reply:
I don't care about the mirrors case. Folks can ignore the lint there. Might be worth adding a note to the description, though.

Reply:
Hrm. We have 2 options.

1. Leave as is and document the part case – maybe push folks to drop named parts.
2. Be safe and just skip this lint for libraries with parts – which are pretty rare.

Otherwise we have to go digging through the analyzer context (or whatever) to find the corresponding part and see if it uses names, etc etc – right?

Reply:
It wouldn't be for all code generators – just code generators that use the old part syntax. Which I think is mostly dead!

Reply:
done!